### PR TITLE
Better FP16 support in pytorch fp16 utils.

### PR DESCRIPTION
**Pull request by jma127**
This commit adds an FP16Model class as a successor to `network_to_half()`.

The benefits of this class are:

- Preservation of single-precision for BatchNorm layers. The models
  generated by `network_to_half()` convert BatchNorm moment tensors to
  half-precision, then back to single-precision, which hurts the
  accuracy of the moment estimators and occasionally results in NaNs.
- Support for multi-argument `nn.Module`s (self-explanatory from code).

### CODE REVIEW

**Comment by jma127** [comment]
cc @ngimel @mcarilli for review

**Comment by ngimel** [comment]
Ah I see, makes sense.

**Comment by jma127** [comment]
@ngimel Any concerns or does this look good to merge?

**Comment by jma127** [comment]
Added a barebones test with a custom dummy net. Probably makes sense to avoid torchvision dependencies if it can be avoided.
